pkg/agent/controller/networkpolicy/reconciler.go

```diff
@@ -350,10 +528,74 @@ func ipNetToOFAddress(in v1beta1.IPNet) *openflow.IPNetAddress {
 	return openflow.NewIPNetAddress(ipNet)
 }
 
-func ipAddressToOFAddress(in v1beta1.IPAddress) *openflow.IPAddress {
-	return openflow.NewIPAddress(net.IP(in))
+func ipsToOFAddresses(ips sets.String) []types.Address {
+	// Must not return nil as it means not restricted by addresses in Openflow implementation.
+	from := make([]types.Address, 0, len(ips))
+	for ip := range ips {
+		from = append(from, openflow.NewIPAddress(net.ParseIP((ip))))
+	}
+	return from
 }
 
-func ipStringToOFAddress(in string) *openflow.IPAddress {
-	return openflow.NewIPAddress(net.ParseIP(in))
+func servicesToNetworkPolicyPort(in []v1beta1.Service) []*networkingv1.NetworkPolicyPort {
+	// Empty or nil slice means allowing all ports in Kubernetes.
+	// nil must be returned to meet ofClient's expectation for this behavior.
+	if len(in) == 0 {
+		return nil
+	}
+	// It makes sure `out` won't be nil, so that even if only named ports are
+	// specified and none of them are resolvable, the rule just falls back to
+	// allowing no port, instead of all ports.
+	out := make([]*networkingv1.NetworkPolicyPort, 0, len(in))
+	for _, s := range in {
+		service := &networkingv1.NetworkPolicyPort{}
+		if s.Protocol != nil {
+			protoStr := string(*s.Protocol)
+			proto := corev1.Protocol(protoStr)
+			service.Protocol = &proto
+		}
+		if s.Port != nil {
+			// All resolvable named port have been converted to intstr.Int,
+			// ignore unresolvable ones.
+			if s.Port.Type == intstr.String {
+				continue
+			}
+			service.Port = s.Port
+		}
+		out = append(out, service)
+	}
+	return out
+}
+
+// resolveService resolves the port name of the provided service to a port number
+// for the provided Pod.
+func resolveService(service *v1beta1.Service, pod *v1beta1.GroupMemberPod) *v1beta1.Service {
+	// If port is not specified or is already a number, return it as is.
+	if service.Port == nil || service.Port.Type == intstr.Int {
+		return service
+	}
+	for _, port := range pod.Ports {
+		if port.Name == service.Port.StrVal && port.Protocol == *service.Protocol {
+			resolvedPort := intstr.FromInt(int(port.Port))
+			return &v1beta1.Service{Protocol: service.Protocol, Port: &resolvedPort}
+		}
+	}
+	klog.Errorf("Can not resolve port %s for Pod %v", service.Port.StrVal, pod)
```

Comment:
is this an Error?

Reply:
Changed to warning, it's at least an user error, a networkpolicy including named port is applied to a Pod that doesn't have the port name.